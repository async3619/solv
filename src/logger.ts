import * as moment from "moment";
import * as chalk from "chalk";

export type LogLevel = "info" | "error" | "warn";
type Func<A, R> = (arg: A) => R;

const colors: Record<LogLevel, Func<string, string>> = {
    info: chalk.blue,
    warn: chalk.yellow,
    error: chalk.red,
};

function log(content: string, level: LogLevel) {
    console.info(`${colors[level](`[${level}]`)}${chalk.green(`[${moment().format("HH:mm:ss.SSS")}]`)} ${content}`);
}

const logger: Record<LogLevel, Func<string, void>> = {
    info: content => log(content, "info"),
    error: content => log(content, "error"),
    warn: content => log(content, "warn"),
};

export default logger;
