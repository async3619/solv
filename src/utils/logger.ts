import * as moment from "moment";
import * as chalk from "chalk";

export type LogLevel = "info" | "error" | "warn";
type Func<A extends any[], R> = (...arg: A) => R;

const colors: Record<LogLevel, Func<[string], string>> = {
    info: chalk.blue,
    warn: chalk.yellow,
    error: chalk.red,
};

function log(content: string, level: LogLevel, time: moment.Moment = moment()) {
    console.info(`${colors[level](`[${level}]`)}${chalk.green(`[${time.format("HH:mm:ss.SSS")}]`)} ${content}`);
}

type LoggerFn = (content: string, time?: moment.Moment) => void;
const logger: Record<LogLevel, LoggerFn> = {
    info: (content, time) => log(content, "info", time),
    error: (content, time) => log(content, "error", time),
    warn: (content, time) => log(content, "warn", time),
};

export default logger;
