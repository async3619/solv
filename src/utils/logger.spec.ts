import * as moment from "moment";
import * as chalk from "chalk";
import logger from "./logger";

describe("logger", () => {
    beforeEach(() => {
        (chalk.level as any) = 3;
    });

    it("should leaves a log depending on levels properly", async () => {
        const buffers: any[] = [];
        console.info = (...data: any[]) => buffers.push(...data);

        logger.info("Test", moment("2022-01-01 00:00:00", "YYYY-MM-DD HH:mm:ss.SSS"));
        logger.error("Test", moment("2022-01-01 00:00:00", "YYYY-MM-DD HH:mm:ss.SSS"));
        logger.warn("Test", moment("2022-01-01 00:00:00", "YYYY-MM-DD HH:mm:ss.SSS"));

        expect(buffers).toMatchSnapshot();
    });

    it("should leaves a log without passing time data", async () => {
        const buffers: string[] = [];
        console.info = (...data: string[]) => buffers.push(...data);

        logger.info("Test");

        expect(buffers[0].includes("Test")).toBeTruthy();
    });
});
