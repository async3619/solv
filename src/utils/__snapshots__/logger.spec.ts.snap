// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger should leaves a log depending on levels properly 1`] = `
[
  "[34m[info][39m[32m[00:00:00.000][39m Test",
  "[31m[error][39m[32m[00:00:00.000][39m Test",
  "[33m[warn][39m[32m[00:00:00.000][39m Test",
]
`;
